pkg/apis/core/validation/execution.go

```diff
@@ -40,21 +30,122 @@ func ValidateExecutionSpec(fldpath *field.Path, spec core.ExecutionSpec) field.E
 		tmplPath := fldPath.Index(i)
 		if len(tmpl.Name) != 0 {
 			if names.Has(tmpl.Name) {
 				allErrs = append(allErrs, field.Duplicate(tmplPath, tmpl.Name))
+				hasDuplicates = true
 			}
 			names.Insert(tmpl.Name)
 			tmplPath = tmplPath.Key(tmpl.Name)
 		}
 		allErrs = append(allErrs, ValidateDeployItemTemplate(tmplPath, tmpl)...)
 	}
 
+	if !hasDuplicates { // cycle check identifies items by name and the behaviour is undefined if duplicate items are present
+		done := sets.NewString()
+		cycles := []Cycle{}
+		undefined := []UndefinedDeployItemReference{}
+		for i := range list {
+			c, u, d := getCyclesAndUndefinedDependencies(list, i, []string{}, done)
+			cycles = append(cycles, c...)
+			undefined = append(undefined, u...)
+			done.Insert(d.UnsortedList()...)
+		}
+		for _, c := range cycles {
+			allErrs = append(allErrs, field.Invalid(fldPath, c, "cycle found in dependencies"))
+		}
+		for _, u := range undefined {
+			idx, _ := getDeployItemTemplateByName(list, u.Source)
+			allErrs = append(allErrs, field.Invalid(fldPath.Index(idx).Key(u.Source), u.Target, "depends on undefined deploy item"))
+		}
+	}
+
 	return allErrs
 }
 
+// takes arguments:
+// 1. a list of deploy item templates
+// 2. the index of the deploy item template that should be validated
+// 3. a list of already visited deploy item templates (referenced by name), shows the current state of the depth-first search
+// 4. a set of deploy item templates (referenced by name) that have already been checked and can be ignored
+// returns:
+// 1. a list of Cycle objects, representing found cyclic dependencies
+// 2. a list of UndefinedDeployItemReference objects, representing found dependencies to undefined deploy items
+// 3. a set of all deploy item templates (referenced by name) that have already been checked (necessary to avoid finding the same cycle multiple times)
+func getCyclesAndUndefinedDependencies(list core.DeployItemTemplateList, index int, visited []string, done sets.String) ([]Cycle, []UndefinedDeployItemReference, sets.String) {
+	current := list[index]
+	cycles := []Cycle{}
+	undefined := []UndefinedDeployItemReference{}
+	if done.Has(current.Name) { // current deploy item has already been handled
+		return cycles, undefined, done
+	}
+	done.Insert(current.Name)
+	if len(current.DependsOn) == 0 { // current deploy item doesn't depend on anything
+		done.Insert(current.Name)
+		return cycles, undefined, done
+	}
+	visitedNew := append(visited, current.Name)
+	for _, do := range current.DependsOn {
+		hasCycle := false
+		for i, e := range visitedNew { // check if current dependency has already been visited
+			if do == e {
+				// cycle found, identify elements of cycle
+				hasCycle = true
+				cycle := []string{}
+				for j := i; j < len(visitedNew); j++ {
+					cycle = append(cycle, visitedNew[j])
+				}
+				cycles = append(cycles, cycle)
+				break
+			}
+		}
+		if hasCycle {
+			continue
+		}
+		idx, _ := getDeployItemTemplateByName(list, do)
```

Comment:
you could use the names set from [here](https://github.com/gardener/landscaper/pull/33/files#diff-726c06f3772623afef302638b764886c7ee7f2a0877c624685ff003d3bf38971R32) to check if the deployitem exists

Reply:
But I would have to provide that names set as an additional argument to the `getCyclesAndUndefinedDependencies` method, right? I don't think that this is a good idea, the method already has more than enough arguments, so I don't like the idea of adding one just for easier access to some information that is already present ...